File: ClassroomTester.java
import java.util.Scanner;
/**
 * UNIT: Objects and Classes
 * 
 * COMPLETED CLASSES:
 * ClassroomTester.java
 * 
 * ASSESSED CLASSES:
 * Classroom.java
 * Student.java
 * 
 * ASSESSMENT ON: 
 * -The ability to manipulate attributes and create methods within a class
 * -The knowledge of how to flush the input stream (i.e. nextLine() after nextInt())
 * 
 * NOT ASSESSED AT ALL:
 * -The knowledge or usage of several library classes, such as javax.swing
 * -The knowledge and understanding of the concept of inheritance
 * -The ability to use classes and inherited classes from client classes
 *         that are outside of the target class's hierarchy
 * 
 */
public class ClassroomTester {
    public static void main(String[] args) {
        System.out.println("Imagine you are a teacher in a classroom. This" +
        " program models the classroom environment, \n including keeping track" +
        " of the number of students in the classroom, their grades, and assigning" +
        " tests to change the grades of each student. \n This classroom is " +
        " created with a default average class grade and starts with no students" +
        " inside the classroom. \n Across the program, you can choose to add new " +
        " students or remove underperforming students from the classroom. \n" +
        " The primary feature of this program is the ability to create Multiple Choice tests " +
        " for every student to fill out. One of these students will fill the test " +
        " out manually, while the rest of the students fill it out automatically. \n" +
        " At the end of the test, the overall grade average for that test will be printed." +
        " \n Have fun!");
        
        Scanner in = new Scanner(System.in);
        Classroom classroom = new Classroom();
        int userInput = 0;
        while (userInput != -1) {
            do {
                System.out.println("What would you like to do with your classroom? (enter number)");
                System.out.println("1) Add new student");
                System.out.println("2) Remove existing student");
                System.out.println("3) Assign a new test");
                System.out.println("4) Get the class average");
                System.out.println("5) Get the list of all students in the classroom");
                System.out.println("(Enter -1 to quit)");
                userInput = in.nextInt();
                in.nextLine();
            } while (userInput != -1 && userInput < 1  || userInput > 5);
            
            if (userInput == 1) {
                System.out.println("What is the name of this new student?");
                String name = in.nextLine();
                System.out.println("How many tests has this student taken?");
                int testsTaken = in.nextInt();
                in.nextLine();
                
                int averageGrade = 0;
                while (averageGrade < 1 || averageGrade > 100) {
                    
                System.out.println("What is this student's grade average? " +
                                    "(enter number between 1-100)");
                                    
                    averageGrade = in.nextInt();
                    in.nextLine();
                }
                
                int totalTestScore = testsTaken * averageGrade;
                    
                classroom.addStudent(name, totalTestScore, testsTaken);
            }
            
            else if (userInput == 2) {
                if (classroom.currentSize == 0)
                    System.out.println("No students to remove");
                else {
                    System.out.println("What is the name of this new student?");
                    String targetName = in.nextLine();
                    
                    classroom.removeStudent(targetName);
                }
            }
            
            else if (userInput == 3) {
                System.out.println("How many questions are on this test?");
                int size = in.nextInt();
                in.nextLine();
                
                int passingGrade = 0;
                while (passingGrade < 1 || passingGrade > 100) {
                    
                System.out.println("What is the passing grade on this test? " +
                                    "(enter number between 1-100)");
                                    
                    passingGrade = in.nextInt();
                    in.nextLine();
                }
                
                classroom.assignTest(size, passingGrade);
            }
            
            else if (userInput == 4) {
                System.out.println("Average class grade: " + classroom.averageGrade);
            }
            else if (userInput == 5) {
                for (int i = 0; i < classroom.currentSize; i++) {
                    System.out.println("Name: " + classroom.students[i].name +
                                        "\t Grade: " + classroom.students[i].returnAverageGrade() );
                }
            }
        }
    }
}

File: ClassroomTester.py
import random


class Student:
    """
      NOTE TO STUDENTS: This class has all attributes and methods completed,
      except for two methods: the returnAverageGrade() method and the
      takeTest() method. Complete both of these methods, and this class
      will be complete.
    ATTRIBUTES OF A STUDENT:
    name: The name of the student
    cumulative_score: The total amount of points this student has received across all tests they've taken
    total_tests_taken: The total number of tests this student has taken
    """

    def __init__(self, name_in, total_score=0, total_tests=0):
        self.name = name_in
        self.cumulative_score = total_score
        self.total_tests_taken = total_tests

    def return_average_grade(self):
        #Return the average grade of the student
        #Hint: The formula to find an average in math is:
        #		(Total sum of all numbers) / (The AMOUNT of numbers)
        return self.cumulative_score / self.total_tests_taken

    def take_test(self, questions, correct_answers, class_grade, is_manual):
        points_gained = 0

        if is_manual:
            #Simulate taking a real given test by looping through each question and asking the user for their answer on each question.
            #Update the amount of points the student gains with each question.
            """
            NOTE: If the student gets an answer wrong, nothing happens. They
            do not gain points, but they should not lose any either.
            """
            for i in range(0, len(questions)):
                user_answer = int(input(questions[i]))
                if user_answer == correct_answers[i]:
                    points_gained += 1

        else:
            #Automates taking a test
            for i in questions:
                correct_chance = random.randint(0, 100)

                #NOTE: If the class grade is higher, it will be EASIER to
                #randomly gain a point correctly.
                if correct_chance <= class_grade:
                    points_gained += 1

        return int(100 * points_gained / len(questions))


class Classroom:
    """
      This class has all attributes and constructors fully complete.
     * You must finish the remaining methods that are not complete.
    ATTRIBUTES OF A CLASSROOM:
    students: A list that contains every student inside the classroom
        average_grade: The class average calculated from the grades of every student
    """

    def __init__(self, average_class_percent=80):
        self.students = []
        self.average_grade = average_class_percent

    def assign_test(self, size, passing_grade):

        #INSTRUCTIONS FOR WRITING THIS METHOD:
        """
         * 1: Create three new lists. Two of these lists have a size of the
         *         "size" parameter and represent each question on the test,
         *         and that question's answer represented by an INTEGER. The 
         *         third list has a size equal to the size of the "students" list 
         *         and represents the grade of each student for this test.
        """
        print(f"CREATING NEW TEST WITH {size} QUESTIONS")
        questions = []
        answers = []
        grades = []
        """
         * 2: [COMPLETE] Use a for loop that iterates through each question on the test.
         *         Inside this loop, ask for what the question states, and what
         *         each answer is. NOTE THAT EVERY QUESTION IS MULTIPLE CHOICE!
         *         This means that for EVERY question, you must provide an answer
         *         for each number, THEN provide the number for the CORRECT number.
         *         Obtain the question and answers from the user input, then
         *         store each question and answer into the corresponding index of
         *         the questions list and answers list.
        """
        #This code is provided for you.
        for i in range(0, size):
            questions.append(
                input("Enter the question for question " + str(i + 1) + ":"))
            for j in range(1, 5):
                ansOption = input("Enter an answer OPTION for question " +
                               str(i + 1) + ", Choice" + str(j) + ": ")
                questions[i] += "\n Choice " + str(j) + ": " + str(ansOption)
            answers.append(
                int(
                    input("Enter the correct answer for question " + str(i + 1) +
                          "(enter a number for 1, 2, 3, or 4): ")))
            while answers[i] < 1 or answers[i] > 4:
                answers[i] = int(
                    input("Enter the correct answer for question " + str(i + 1) +
                          "(enter a number for 1, 2, 3, or 4): "))
        """
         * 3: Create another for loop that iterates through each STUDENT
         *         in the classroom.
         *             
         *         This for loop will make each of them take
         *         the test, as seen in the Student class's function. 
         * 
         *         You may choose how many students automate the test and how 
         *         many students manually take the test. However, it is highly 
         *         recommended that ONLY ONE student manually takes the test, 
         *         and every other student will automate the test. 
         * 
         *         You can control who manually and automatically takes the test 
         *         by passing in either a true(manual) or false(automatic) 
         *         boolean variable into the takeTest method, 
         *         as described in the Student class.
        """
        grades.append(self.students[0].take_test(questions, answers, average_grade, True))
        for i in range(1, len(self.students)):
            grades.append(self.students[i].take_test(questions, answers, average_grade, False))
        """
         * 4: Print the grades that each student got on the test, and print
         *         which students have passed or not passed based on if their
         *         test grade meets the passing grade.
        """
        print("\nTEST RESULTS")
        for i in range(0, len(self.students)):
            print(self.students[i].name, "got a ", str(grades[i]), "on the test!")
            if grades[i] > passing_grade:
                print("They pass!")
            else:
                print("Unfortunately, they did not pass.")
    
        # 5: Calculate and print the average test score
        total_grade = 0;
        for grade in grades:
            total_grade += grade;
        print("AVERAGE CLASS GRADE:", str(total_grade / len(grades)))

        # 6: Update the average class grade
        self.average_grade = total_grade / len(grades)

    def add_student(self, name, total_test_score=0, num_tests=0):
        self.students.append(Student(name, total_test_score, num_tests))

    def remove_student(self, name):
        found_student = False
        for student in self.students:
            if student.name == name:
                found_student = True
                self.students.remove(student)
                break

        if not found_student:
            print("Unable to find student with name ", name)


print(
    "Imagine you are a teacher in a classroom. This program models the classroom environment, "
    "including keeping track of the number of students in the classroom, their grades, and assigning "
    "tests to change the grades of each student. This classroom is created with a default average class "
    "grade and starts with no students inside the classroom. Across the program, you can choose to add new "
    "students or remove underperforming students from the classroom. The primary feature of this program is "
    "the ability to create Multiple Choice tests for every student to fill out. One of these students will "
    "fill the test out manually, while the rest of the students fill it out automatically. At the end of the "
    "test, the overall grade average for that test will be printed. Have fun!")

classroom = Classroom()
user_input = 0
while user_input != -1:
    user_input = int(
        input(
            "What would you like to do with your classroom? (enter number)\n1) Add new student \n2) Remove existing student \n3) Assign a new test \n4) Get the class average \n5) Get the list of all students in the classroom \n(Enter -1 to quit): "
        ))

    if user_input == 1:
        name = input("What is the name of this new student? ")
        tests_taken = int(input("How many tests has this student taken? "))
        average_grade = 0
        while average_grade < 1 or average_grade > 100:
            average_grade = int(
                input(
                    "What is this student's grade average? (enter number between 1-100): "
                ))
        total_test_score = tests_taken * average_grade
        classroom.add_student(name, total_test_score, tests_taken)

    elif user_input == 2:
        if len(classroom.students) == 0:
            print("No students to remove")
        else:
            target_name = input("What is the name of the student to remove? ")
            classroom.remove_student(target_name)

    elif user_input == 3:
        size = int(input("How many questions are on this test? "))
        passing_grade = 0
        while passing_grade < 1 or passing_grade > 100:
            passing_grade = int(
                input(
                    "What is the passing grade on this test? (enter number between 1-100): "
                ))
        classroom.assign_test(size, passing_grade)

    elif user_input == 4:
        print(f"Average class grade: {classroom.average_grade}")

    elif user_input == 5:
        for student in classroom.students:
            if student is not None:
                print(
                    f"Name: {student.name}\t Grade: {student.return_average_grade()}"
                )
